File: deepdoc/vision/ocr_tiling.py
#
#  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

from collections import defaultdict
from collections.abc import Callable
from itertools import product

import numpy as np

DEFAULT_TILE_SIZE = 2880
DEFAULT_TILE_OVERLAP = 288
DEFAULT_TILING_THRESHOLD = 4096
DEFAULT_DUPLICATE_OVERLAP = 0.5


def tile_starts(length: int, tile_size: int, overlap: int) -> list[int]:
    """Return full-size tile starts that cover an axis, including its end."""
    if tile_size <= 0:
        raise ValueError("tile_size must be positive")
    if overlap < 0 or overlap >= tile_size:
        raise ValueError("overlap must be non-negative and smaller than tile_size")
    if length <= tile_size:
        return [0]

    step = tile_size - overlap
    starts = list(range(0, length - tile_size + 1, step))
    final_start = length - tile_size
    if starts[-1] != final_start:
        starts.append(final_start)
    return starts


def _box_bounds(box: np.ndarray) -> tuple[float, float, float, float]:
    """Return an axis-aligned bounding rectangle for a quadrilateral."""
    return float(np.min(box[:, 0])), float(np.min(box[:, 1])), float(np.max(box[:, 0])), float(np.max(box[:, 1]))


def _intersection_over_smaller(bounds_a: tuple[float, float, float, float], bounds_b: tuple[float, float, float, float]) -> float:
    """Measure intersection relative to the smaller rectangle."""
    left = max(bounds_a[0], bounds_b[0])
    top = max(bounds_a[1], bounds_b[1])
    right = min(bounds_a[2], bounds_b[2])
    bottom = min(bounds_a[3], bounds_b[3])
    intersection = max(0.0, right - left) * max(0.0, bottom - top)
    area_a = max(0.0, bounds_a[2] - bounds_a[0]) * max(0.0, bounds_a[3] - bounds_a[1])
    area_b = max(0.0, bounds_b[2] - bounds_b[0]) * max(0.0, bounds_b[3] - bounds_b[1])
    smaller_area = min(area_a, area_b)
    return intersection / smaller_area if smaller_area else 0.0


def _axis_aligned_box(bounds: tuple[float, float, float, float]) -> np.ndarray:
    """Convert rectangle bounds to a clockwise quadrilateral."""
    left, top, right, bottom = bounds
    return np.asarray([[left, top], [right, top], [right, bottom], [left, bottom]], dtype=np.float32)


def deduplicate_boxes(
    boxes: list[np.ndarray],
    sources: list[int] | None = None,
    overlap_threshold: float = DEFAULT_DUPLICATE_OVERLAP,
    cell_size: int = DEFAULT_TILE_OVERLAP,
) -> list[np.ndarray]:
    """Merge duplicate detections created by overlapping tiles."""
    if not boxes:
        return []
    if sources is None:
        sources = list(range(len(boxes)))
    if len(sources) != len(boxes):
        raise ValueError("sources and boxes must have the same length")

    bounds = [_box_bounds(box) for box in boxes]
    areas = [(right - left) * (bottom - top) for left, top, right, bottom in bounds]
    candidate_indices = sorted(range(len(boxes)), key=lambda index: areas[index], reverse=True)
    spatial_index: dict[tuple[int, int], list[int]] = defaultdict(list)
    kept_boxes: list[np.ndarray] = []
    kept_bounds: list[tuple[float, float, float, float]] = []
    kept_sources: list[set[int]] = []
    active: list[bool] = []

    for candidate_index in candidate_indices:
        candidate_bounds = bounds[candidate_index]
        left_cell = int(candidate_bounds[0] // cell_size)
        top_cell = int(candidate_bounds[1] // cell_size)
        right_cell = int(candidate_bounds[2] // cell_size)
        bottom_cell = int(candidate_bounds[3] // cell_size)
        cells = [(x, y) for x in range(left_cell, right_cell + 1) for y in range(top_cell, bottom_cell + 1)]
        nearby = {index for cell in cells for index in spatial_index[cell] if active[index]}

        duplicates = [index for index in nearby if sources[candidate_index] not in kept_sources[index] and _intersection_over_smaller(candidate_bounds, kept_bounds[index]) >= overlap_threshold]
        if duplicates:
            duplicate_index = min(duplicates)
            merged_bounds = (
                min(candidate_bounds[0], *(kept_bounds[index][0] for index in duplicates)),
                min(candidate_bounds[1], *(kept_bounds[index][1] for index in duplicates)),
                max(candidate_bounds[2], *(kept_bounds[index][2] for index in duplicates)),
                max(candidate_bounds[3], *(kept_bounds[index][3] for index in duplicates)),
            )
            if merged_bounds != kept_bounds[duplicate_index]:
                kept_boxes[duplicate_index] = _axis_aligned_box(merged_bounds)
            kept_bounds[duplicate_index] = merged_bounds
            kept_sources[duplicate_index] = {sources[candidate_index]}.union(*(kept_sources[index] for index in duplicates))
            for index in duplicates:
                if index != duplicate_index:
                    active[index] = False

            merged_left_cell = int(merged_bounds[0] // cell_size)
            merged_top_cell = int(merged_bounds[1] // cell_size)
            merged_right_cell = int(merged_bounds[2] // cell_size)
            merged_bottom_cell = int(merged_bounds[3] // cell_size)
            for x in range(merged_left_cell, merged_right_cell + 1):
                for y in range(merged_top_cell, merged_bottom_cell + 1):
                    if duplicate_index not in spatial_index[(x, y)]:
                        spatial_index[(x, y)].append(duplicate_index)
            continue

        kept_index = len(kept_boxes)
        kept_boxes.append(boxes[candidate_index])
        kept_bounds.append(candidate_bounds)
        kept_sources.append({sources[candidate_index]})
        active.append(True)
        for cell in cells:
            spatial_index[cell].append(kept_index)

    return [box for index, box in enumerate(kept_boxes) if active[index]]


def detect_tiled_boxes(
    image: np.ndarray,
    detect_tile: Callable[[np.ndarray], np.ndarray | None],
    tile_size: int = DEFAULT_TILE_SIZE,
    overlap: int = DEFAULT_TILE_OVERLAP,
) -> np.ndarray:
    """Detect text in overlapping tiles and map the boxes to image coordinates."""
    height, width = image.shape[:2]
    boxes: list[np.ndarray] = []
    sources: list[int] = []

    tile_origins = product(tile_starts(height, tile_size, overlap), tile_starts(width, tile_size, overlap))
    for tile_index, (top, left) in enumerate(tile_origins):
        tile_boxes = detect_tile(image[top : top + tile_size, left : left + tile_size])
        if tile_boxes is None:
            continue
        for box in tile_boxes:
            global_box = np.asarray(box, dtype=np.float32).copy()
            global_box[:, 0] += left
            global_box[:, 1] += top
            boxes.append(global_box)
            sources.append(tile_index)

    deduplicated = deduplicate_boxes(boxes, sources=sources, cell_size=max(overlap, 1))
    if not deduplicated:
        return np.empty((0, 4, 2), dtype=np.float32)
    return np.asarray(deduplicated, dtype=np.float32)

File: internal/deepdoc/parser/pdf/inference/ocr_tiling.go
package inference

import (
	"image"
	"math"
	"sort"

	pdf "ragflow/internal/deepdoc/parser/pdf/type"
	util "ragflow/internal/deepdoc/parser/pdf/util"
)

// A 4096px threshold keeps common 300-DPI document pages on one request.
// Larger images would be reduced below 0.234x by the detector's 960px input;
// 2880px tiles retain about 0.33x scale, with a 10% overlap for edge text.
const (
	ocrTileSize         = 2880
	ocrTileOverlap      = 288
	ocrTilingThreshold  = 4096
	ocrDuplicateOverlap = 0.5
)

type ocrBoxBounds struct {
	left, top, right, bottom float64
}

type ocrGridCell struct {
	x, y int
}

func ocrTileStarts(length int) []int {
	if length <= ocrTileSize {
		return []int{0}
	}
	step := ocrTileSize - ocrTileOverlap
	starts := make([]int, 0, (length+step-1)/step)
	for start := 0; start <= length-ocrTileSize; start += step {
		starts = append(starts, start)
	}
	finalStart := length - ocrTileSize
	if starts[len(starts)-1] != finalStart {
		starts = append(starts, finalStart)
	}
	return starts
}

func detectTiledOCR(img image.Image, detectTile func(image.Image) ([]pdf.OCRBox, error)) ([]pdf.OCRBox, error) {
	bounds := img.Bounds()
	rowStarts := ocrTileStarts(bounds.Dy())
	columnStarts := ocrTileStarts(bounds.Dx())
	boxes := make([]pdf.OCRBox, 0)
	sources := make([]int, 0)
	tileIndex := 0
	for _, top := range rowStarts {
		for _, left := range columnStarts {
			tile := util.FastCrop(
				img,
				bounds.Min.X+left,
				bounds.Min.Y+top,
				bounds.Min.X+min(left+ocrTileSize, bounds.Dx()),
				bounds.Min.Y+min(top+ocrTileSize, bounds.Dy()),
			)
			tileBoxes, err := detectTile(tile)
			if err != nil {
				return nil, err
			}
			for _, box := range tileBoxes {
				boxes = append(boxes, translateOCRBox(box, float64(left), float64(top)))
				sources = append(sources, tileIndex)
			}
			tileIndex++
		}
	}
	return deduplicateOCRBoxes(boxes, sources), nil
}

func translateOCRBox(box pdf.OCRBox, left, top float64) pdf.OCRBox {
	box.X0 += left
	box.X1 += left
	box.X2 += left
	box.X3 += left
	box.Y0 += top
	box.Y1 += top
	box.Y2 += top
	box.Y3 += top
	return box
}

func boundsForOCRBox(box pdf.OCRBox) ocrBoxBounds {
	return ocrBoxBounds{
		left:   min(box.X0, box.X1, box.X2, box.X3),
		top:    min(box.Y0, box.Y1, box.Y2, box.Y3),
		right:  max(box.X0, box.X1, box.X2, box.X3),
		bottom: max(box.Y0, box.Y1, box.Y2, box.Y3),
	}
}

func (b ocrBoxBounds) area() float64 {
	return max(0, b.right-b.left) * max(0, b.bottom-b.top)
}

func (b ocrBoxBounds) cells() []ocrGridCell {
	left := int(math.Floor(b.left / ocrTileOverlap))
	top := int(math.Floor(b.top / ocrTileOverlap))
	right := int(math.Floor(b.right / ocrTileOverlap))
	bottom := int(math.Floor(b.bottom / ocrTileOverlap))
	cells := make([]ocrGridCell, 0, (right-left+1)*(bottom-top+1))
	for x := left; x <= right; x++ {
		for y := top; y <= bottom; y++ {
			cells = append(cells, ocrGridCell{x: x, y: y})
		}
	}
	return cells
}

func intersectionOverSmaller(a, b ocrBoxBounds) float64 {
	left := max(a.left, b.left)
	top := max(a.top, b.top)
	right := min(a.right, b.right)
	bottom := min(a.bottom, b.bottom)
	intersection := max(0, right-left) * max(0, bottom-top)
	smallerArea := min(a.area(), b.area())
	if smallerArea == 0 {
		return 0
	}
	return intersection / smallerArea
}

func axisAlignedOCRBox(bounds ocrBoxBounds) pdf.OCRBox {
	return pdf.OCRBox{
		X0: bounds.left, Y0: bounds.top,
		X1: bounds.right, Y1: bounds.top,
		X2: bounds.right, Y2: bounds.bottom,
		X3: bounds.left, Y3: bounds.bottom,
	}
}

func deduplicateOCRBoxes(boxes []pdf.OCRBox, sources []int) []pdf.OCRBox {
	if len(boxes) == 0 {
		return nil
	}
	if len(sources) != len(boxes) {
		return boxes
	}

	bounds := make([]ocrBoxBounds, len(boxes))
	order := make([]int, len(boxes))
	for i, box := range boxes {
		bounds[i] = boundsForOCRBox(box)
		order[i] = i
	}
	sort.SliceStable(order, func(i, j int) bool {
		return bounds[order[i]].area() > bounds[order[j]].area()
	})

	spatialIndex := make(map[ocrGridCell][]int)
	keptBoxes := make([]pdf.OCRBox, 0, len(boxes))
	keptBounds := make([]ocrBoxBounds, 0, len(boxes))
	keptSources := make([]map[int]struct{}, 0, len(boxes))
	active := make([]bool, 0, len(boxes))

	for _, candidateIndex := range order {
		candidateBounds := bounds[candidateIndex]
		nearby := make(map[int]struct{})
		for _, cell := range candidateBounds.cells() {
			for _, index := range spatialIndex[cell] {
				if active[index] {
					nearby[index] = struct{}{}
				}
			}
		}

		duplicates := make([]int, 0)
		for index := range nearby {
			if _, sameSource := keptSources[index][sources[candidateIndex]]; sameSource {
				continue
			}
			if intersectionOverSmaller(candidateBounds, keptBounds[index]) >= ocrDuplicateOverlap {
				duplicates = append(duplicates, index)
			}
		}
		if len(duplicates) > 0 {
			sort.Ints(duplicates)
			duplicateIndex := duplicates[0]
			mergedBounds := candidateBounds
			mergedSources := map[int]struct{}{sources[candidateIndex]: {}}
			for _, index := range duplicates {
				mergedBounds.left = min(mergedBounds.left, keptBounds[index].left)
				mergedBounds.top = min(mergedBounds.top, keptBounds[index].top)
				mergedBounds.right = max(mergedBounds.right, keptBounds[index].right)
				mergedBounds.bottom = max(mergedBounds.bottom, keptBounds[index].bottom)
				for source := range keptSources[index] {
					mergedSources[source] = struct{}{}
				}
				if index != duplicateIndex {
					active[index] = false
				}
			}
			if mergedBounds != keptBounds[duplicateIndex] {
				keptBoxes[duplicateIndex] = axisAlignedOCRBox(mergedBounds)
			}
			keptBounds[duplicateIndex] = mergedBounds
			keptSources[duplicateIndex] = mergedSources
			for _, cell := range mergedBounds.cells() {
				alreadyIndexed := false
				for _, index := range spatialIndex[cell] {
					if index == duplicateIndex {
						alreadyIndexed = true
						break
					}
				}
				if !alreadyIndexed {
					spatialIndex[cell] = append(spatialIndex[cell], duplicateIndex)
				}
			}
			continue
		}

		keptIndex := len(keptBoxes)
		keptBoxes = append(keptBoxes, boxes[candidateIndex])
		keptBounds = append(keptBounds, candidateBounds)
		keptSources = append(keptSources, map[int]struct{}{sources[candidateIndex]: {}})
		active = append(active, true)
		for _, cell := range candidateBounds.cells() {
			spatialIndex[cell] = append(spatialIndex[cell], keptIndex)
		}
	}

	result := make([]pdf.OCRBox, 0, len(keptBoxes))
	for index, box := range keptBoxes {
		if active[index] {
			result = append(result, box)
		}
	}
	return result
}
